presto-geospatial/src/main/java/com/facebook/presto/geospatial/GeoFunctions.java

```diff
@@ -0,0 +1,513 @@
+    }
+
+    @Description("Returns the last point of a LINESTRING geometry as a Point")
+    @ScalarFunction("ST_EndPoint")
+    @SqlType(NAME)
+    public static Slice stEndPoint(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_EndPoint", LINE_STRING);
+        MultiPath lines = (MultiPath) geometry.getEsriGeometry();
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        return serialize(OGCGeometry.createFromEsriGeometry(lines.getPoint(lines.getPointCount() - 1), reference));
+    }
+
+    @Description("Return the X coordinate of the point")
+    @ScalarFunction("ST_X")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stX(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_X", POINT);
+        return ((OGCPoint) geometry).X();
+    }
+
+    @Description("Return the Y coordinate of the point")
+    @ScalarFunction("ST_Y")
+    @SqlType(StandardTypes.DOUBLE)
+    public static double stY(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = getGeometry(deserialize(input), "ST_Y", POINT);
+        return ((OGCPoint) geometry).Y();
+    }
+
+    @Description("Returns the closure of the combinatorial boundary of this Geometry")
+    @ScalarFunction("ST_Boundary")
+    @SqlType(NAME)
+    public static Slice stBoundary(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        OGCGeometry bound = geometry.boundary();
+        if (bound.geometryType().equals(MULTI_LINE_STRING) && ((OGCMultiLineString) bound).numGeometries() == 1) {
+            bound = ((OGCMultiLineString) bound).geometryN(0);
+        }
+        return serialize(bound);
+    }
+
+    @Description("Returns a geometry representing the double precision (float8) bounding box of the supplied geometry")
+    @ScalarFunction("ST_Envelope")
+    @SqlType(NAME)
+    public static Slice stEnvelope(@SqlType(NAME) Slice input)
+    {
+        OGCGeometry geometry = deserialize(input);
+        SpatialReference reference = geometry.getEsriSpatialReference();
+        Envelope envelope = getEnvelope(geometry);
+        return serialize(OGCGeometry.createFromEsriGeometry(envelope, reference));
+    }
+
+    @Description("Returns a geometry that represents that part of geometry A that does not intersect with geometry B")
+    @ScalarFunction("ST_Difference")
+    @SqlType(NAME)
+    public static Slice stDifference(@SqlType(NAME) Slice left, @SqlType(NAME) Slice right)
+    {
+        OGCGeometry leftGeometry = deserialize(left);
+        OGCGeometry rightGeometry = deserialize(right);
+        return serialize(leftGeometry.difference(rightGeometry));
+    }
+
+    @Description("For geometry type Returns the 2-dimensional cartesian minimum distance (based on spatial ref) between two geometries in projected units")
```

Comment:
nit: "For geometry type" is not needed

The mentioning of "(based on spatial ref)" and "in projected units" comes as a surprise here given that st_area and st_length don't include similar language. Is there a way to update documentation for all 3 functions to make it consistent?

All functions which take two geometries have an implicit requirement for both geometries to have the same spatial reference. However, this condition is never checked explicitly. Although, in this version we can never use geometries with spatial references other than null, it would be helpful to check that SRs of input geometries match in case we allow for non-null SRs in the future.